saleor/checkout/fetch.py

```diff
@@ -147,35 +175,61 @@ def fetch_checkout_lines(checkout: "Checkout") -> Iterable[CheckoutLineInfo]:
             # discount from voucher
             return lines_info
         if voucher.type == VoucherType.SPECIFIC_PRODUCT or voucher.apply_once_per_order:
-            discounted_lines_by_voucher: List[CheckoutLineInfo] = []
-            if voucher.apply_once_per_order:
-                discounts = fetch_active_discounts()
-                channel = checkout.channel
-                cheapest_line_price = None
-                cheapest_line = None
-                for line_info in lines_info:
-                    line_price = line_info.variant.get_price(
-                        product=line_info.product,
-                        collections=line_info.collections,
-                        channel=channel,
-                        channel_listing=line_info.channel_listing,
-                        discounts=discounts,
-                    )
-                    if not cheapest_line or cheapest_line_price > line_price:
-                        cheapest_line_price = line_price
-                        cheapest_line = line_info
-                if cheapest_line:
-                    discounted_lines_by_voucher.append(cheapest_line)
-            else:
-                discounted_lines_by_voucher.extend(
-                    get_discounted_lines(lines_info, voucher)
-                )
-            for line_info in lines_info:
-                if line_info in discounted_lines_by_voucher:
-                    line_info.voucher = voucher
+            _apply_voucher_on_product(voucher, checkout, lines_info)
     return lines_info
 
 
+def _get_variant_channel_listing(variant: "ProductVariant", channel_id: int):
+    variant_channel_listing = None
+    for channel_listing in variant.channel_listings.all():
+        if channel_listing.channel_id == channel_id:
+            variant_channel_listing = channel_listing
+    return variant_channel_listing
+
+
+def _get_product_channel_listing(
+    product_channel_listing_mapping: dict, channel_id: int, product: "Product"
+):
+    product_channel_listing = product_channel_listing_mapping.get(product.id)
+    if product.id not in product_channel_listing_mapping:
+        for channel_listing in product.channel_listings.all():  # type: ignore
+            if channel_listing.channel_id == channel_id:
+                product_channel_listing = channel_listing
+        product_channel_listing_mapping[product.id] = product_channel_listing
+    return product_channel_listing
+
+
+def _apply_voucher_on_product(
+    voucher: "Voucher", checkout: "Checkout", lines_info: Iterable[CheckoutLineInfo]
+):
+    from .utils import get_discounted_lines
```

Comment:
Is any reason for local import here? I mean circular import etc?

Reply:
Yes, circular import is the reason.